homeassistant/components/stream/__init__.py

```diff
@@ -95,6 +100,96 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+class StreamClientError(IntEnum):
+    """Enum for stream client errors."""
+
+    BadRequest = 400
+    Unauthorized = 401
+    Forbidden = 403
+    NotFound = 404
+    Other = 4
+
+
+class StreamOpenClientError(HomeAssistantError):
+    """Raised when client error received when trying to open a stream.
+
+    :param stream_client_error: The type of client error
+    """
+
+    def __init__(
+        self, *args: Any, stream_client_error: StreamClientError, **kwargs: Any
+    ) -> None:
+        self.stream_client_error = stream_client_error
+        super().__init__(*args, **kwargs)
+
+
+async def _async_try_open_stream(
+    hass: HomeAssistant, source: str, pyav_options: dict[str, str] | None = None
+) -> InputContainer | OutputContainer:
+    """Try to open a stream.
+
+    Will raise StreamOpenClientError if an http client error is encountered.
+    """
+    import av  # pylint: disable=import-outside-toplevel
+
+    if not pyav_options:
+        pyav_options = {}
+
+    default_pyav_options = {
+        "rtsp_flags": CONF_PREFER_TCP,
+        "timeout": str(SOURCE_TIMEOUT),
+    }
+
+    pyav_options = {
+        **default_pyav_options,
+        **pyav_options,
+    }
+
+    try:
+        func = partial(av.open, source, options=pyav_options, timeout=5)
+        container = await hass.loop.run_in_executor(None, func)
+
+    except av.HTTPBadRequestError as ex:
+        raise StreamOpenClientError(
+            stream_client_error=StreamClientError.BadRequest
+        ) from ex
+
+    except av.HTTPUnauthorizedError as ex:
+        raise StreamOpenClientError(
+            stream_client_error=StreamClientError.Unauthorized
+        ) from ex
+
+    except av.HTTPForbiddenError as ex:
+        raise StreamOpenClientError(
+            stream_client_error=StreamClientError.Forbidden
+        ) from ex
+
+    except av.HTTPNotFoundError as ex:
+        raise StreamOpenClientError(
+            stream_client_error=StreamClientError.NotFound
+        ) from ex
+
+    except av.HTTPOtherClientError as ex:
+        raise StreamOpenClientError(stream_client_error=StreamClientError.Other) from ex
+
+    else:
+        return container
+
+
+async def async_check_stream_client_error(
+    hass: HomeAssistant, source: str, pyav_options: dict[str, str] | None = None
+) -> StreamClientError | None:
+    """Return an enum value representing a stream client error or None."""
+    try:
+        container = await _async_try_open_stream(hass, source, pyav_options)
+    except StreamOpenClientError as ex:
+        return ex.stream_client_error
+    else:
+        await hass.loop.run_in_executor(None, container.close)
```

Comment:
We should close in the executor instead of doing two executor jobs.

I think `_async_try_open_stream` can become `_try_open_stream` and fully be run in the executor.

Reply:
That's done. Both methods required a sync and async version as the check method closes the stream and the `_async_try_open_stream` doesn't